Assets/MixedRealityToolkit.Services/InputAnimation/Editor/InputRecordingWindow.cs

```diff
@@ -0,0 +1,330 @@
+                    stepFwd = GUILayout.Button(new GUIContent(iconStepFwd, "Step forward one frame"), "Button");
+                }
+
+                float time = PlaybackService.LocalTime;
+                float duration = (animation != null ? animation.Duration : 0.0f);
+                float newTime = GUILayout.HorizontalSlider(time, 0.0f, duration);
+
+                if (play != wasPlaying)
+                {
+                    if (play)
+                    {
+                        PlaybackService.Play();
+                    }
+                    else
+                    {
+                        PlaybackService.Pause();
+                    }
+                }
+                if (jumpBack)
+                {
+                    PlaybackService.LocalTime = 0.0f;
+                }
+                if (stepFwd)
+                {
+                    PlaybackService.LocalTime += Time.deltaTime;
+                }
+                if (newTime != time)
+                {
+                    PlaybackService.LocalTime = newTime;
+                }
+
+                // Repaint while playing to update the timeline
+                if (PlaybackService.IsPlaying)
+                {
+                    Repaint();
+                }
+            }
+        }
+
+        private void DrawAnimationInfo()
+        {
+            using (new GUILayout.VerticalScope(EditorStyles.helpBox))
+            {
+                GUILayout.Label("Animation Info:", EditorStyles.boldLabel);
+
+                if (animation != null)
+                {
+                    GUILayout.Label($"File Path: {loadedFilePath}");
+                    GUILayout.Label($"Duration: {animation.Duration}");
+                }
+            }
+        }
+
+        private void SaveAnimation(bool loadAfterExport)
+        {
+            string outputPath;
+            if (loadedFilePath.Length > 0)
+            {
+                string loadedDirectory = Path.GetDirectoryName(loadedFilePath);
+                outputPath = EditorUtility.SaveFilePanel(
+                    "Select output path",
+                    loadedDirectory,
+                    InputAnimationSerializationUtils.GetOutputFilename(),
+                    InputAnimationSerializationUtils.Extension);
+            }
+            else
+            {
+                outputPath = EditorUtility.SaveFilePanelInProject(
+                    "Select output path",
+                    InputAnimationSerializationUtils.GetOutputFilename(),
+                    InputAnimationSerializationUtils.Extension,
+                    "Enter filename for exporting input animation");
+            }
+
+            if (outputPath.Length > 0)
+            {
+                string filename = Path.GetFileName(outputPath);
+                string directory = Path.GetDirectoryName(outputPath);
+
+                string result = RecordingService.SaveInputAnimation(filename, directory);
```

Comment:
Seeing as though there was a check added to the recording service (that a save would only occur if recording wasn't happening) - would it be worth adding something similar here or something?

More generally it does seem like an implementation detail/contract of the recording service (save should only be supported while recording has stopped), so maybe this should then just be an overload of the recording service itself

Reply:
The "isRecording" check was actually added in the InputRecordingControls script for the demo files. The recording service does not care about when the recorded data is serialized.

The reason for checking the recording state is that the SaveRecordedInput function is triggered by the button click - but it's supposed to work as a toggle button (for which we don't have a true concept in MRTK afaik) and so we get a click event both when starting and stopping the recording. Saving on Start does not make any sense (the buffer would be empty), so i added that check to only record after stopping.